cmd/ops_agent_uap_plugin/service_linux.go

```
@@ -0,0 +1,205 @@
+)
+
+const (
+	OpsAgentConfigLocationLinux = "/etc/google-cloud-ops-agent/config.yaml"
+	ConfGeneratorBinary         = "libexec/google_cloud_ops_agent_engine"
+	LogsDirectory               = "log/google-cloud-ops-agent"
+	FluentBitStateDiectory      = "state/fluent-bit"
+	FluentBitRuntimeDirectory   = "run/google-cloud-ops-agent-fluent-bit"
+	OtelRuntimeDirectory        = "run/google-cloud-ops-agent-opentelemetry-collector"
+	DefaultPluginStateDirectory = "/var/lib/google-guest-agent/plugins/ops-agent-plugin"
+)
+
+var (
+	AgentServiceNameRegex    = regexp.MustCompile(`[\w-]+\.service`)
+	AgentSystemdServiceNames = []string{"google-cloud-ops-agent.service", "stackdriver-agent.service", "google-fluentd.service"}
+)
+
+// Apply applies the config sent or performs the work defined in the message.
+// ApplyRequest is opaque to the agent and is expected to be well known contract
+// between Plugin and the server itself. For e.g. service might want to update
+// plugin config to enable/disable feature here plugins can react to such requests.
+func (ps *OpsAgentPluginServer) Apply(ctx context.Context, msg *pb.ApplyRequest) (*pb.ApplyResponse, error) {
+	return &pb.ApplyResponse{}, nil
+}
+
+// Start starts the plugin and initiates the plugin functionality.
+// Until plugin receives Start request plugin is expected to be not functioning
+// and just listening on the address handed off waiting for the request.
+func (ps *OpsAgentPluginServer) Start(ctx context.Context, msg *pb.StartRequest) (*pb.StartResponse, error) {
+	if ps.cancel != nil {
+		log.Printf("The Ops Agent plugin is started already, skipping the current request")
+		return &pb.StartResponse{}, nil
+	}
+	log.Printf("Received a Start request: %s. Starting the Ops Agent", msg)
+
+	pContext, cancel := context.WithCancel(context.Background())
+	ps.cancel = cancel
+
+	pluginStateDir := msg.GetConfig().GetStateDirectoryPath()
+	if pluginStateDir == "" {
```

Under what conditions will this be empty? If empty is unexpected, we should return an error. On the other hand, if we're required to handle empty, then we should use a default state path that _we_ own, not one that the plugin owns.